### ISSUE TITLE
keyword scrapper container shutdown

### ISSUE THREAD

**Issue by prravda** [prravda/discord-hotdeal-scrapper-typescript#42]
# Status
- Sometime scrapper container shut down.
- Find the root cause and solve it based on log, and metric.

**Comment by prravda** [comment]
# Progresses
## Log checking
```shell
2023-08-19 23:00:41.285	
/app/src/scrappers/fmkorea-hot-deal-scrapper.ts:364

2023-08-19 23:00:41.285	
                const header = await req.allHeaders();
2023-08-19 23:00:41.285	
                                         ^
2023-08-19 23:00:41.298	
request.allHeaders: Target page, context or browser has been closed
2023-08-19 23:00:41.299	
    at Page.<anonymous> (/app/src/scrappers/fmkorea-hot-deal-scrapper.ts:364:42) {
2023-08-19 23:00:41.299	
  name: 'Error'
2023-08-19 23:00:41.299	
}
```
- This is error's stack, but I can not found any logical misss in this logic itself.
- This is the [logic](https://github.com/prravda/discord-hotdeal-scrapper-typescript/blob/main/src/scrappers/fmkorea-hot-deal-scrapper.ts#L370). But the function that running this code is not triggered with input, it uses event emitter. 

## Profiling 

### Configuring and setting-up
- So, I added profiler to watch CPU and Memory load. 
- I used [`pyroscope`](https://pyroscope.io/) because it is easy to configure and deploy using container runtime. 
- The node.js application in this container send it's profile to pyroscope server. 
  - As I mentioned, it is easy to configure. Just installed the library, imported it, and configure with a small code [like this](https://github.com/prravda/discord-hotdeal-scrapper-typescript/blob/main/app.ts#L8C17-L8C17)
- There are other parts for configure and deploy the pyroscope 
  - [pyroscope server configuration .yaml file](https://github.com/prravda/discord-hotdeal-scrapper-typescript/blob/main/grafana/pyroscope-local-config.yaml)
  - [pyroscope section for grafana datasource definition file](https://github.com/prravda/discord-hotdeal-scrapper-typescript/blob/main/grafana/grafana-datasource.yaml#L12) 
## Conclusion and figure out the solution based on this
  - [docker-compose.yaml file for pyroscope deployment](https://github.com/prravda/discord-hotdeal-scrapper-typescript/blob/main/docker-compose.yaml#L37C4-L37C4) 

### Watch the profiling result 

![pyroscope log on grafana](https://github.com/prravda/discord-hotdeal-scrapper-typescript/assets/64535895/81de4304-9952-4d42-80f2-73bb54cdf0b1)
- I collect the profiles for almost a day. And this is the result.

![occupated memories are increasing by time](https://github.com/prravda/discord-hotdeal-scrapper-typescript/assets/64535895/f78fda47-46e7-4c47-a978-4581ef87db68)
- After modifying the graph, the ouccupated memory increasing seems more explicitly.

![profile information - 3 hours after profiling](https://github.com/prravda/discord-hotdeal-scrapper-typescript/assets/64535895/12f5d1e6-a159-46fb-adf0-8f54d1a0c8e6)
- This is the flame graph of memory based on profile information after 3 hours from the starting point of collecting.


![profile information - 12 hours after profiling](https://github.com/prravda/discord-hotdeal-scrapper-typescript/assets/64535895/d3fbbde1-9464-479b-9753-9d6052e25032)
- But after 9 hours, the result changed significantly. The point is memory occupation of `playwright` (headless browser) library is increased.
- It increased from 15.28% to 35.85%.

###Watch the processes list in this container
```shell
root@hotdeal-scrapper:/app# ps -e
    PID TTY          TIME CMD
      1 ?        00:00:01 npm exec ts-nod
     16 ?        00:00:00 sh
     17 ?        00:35:57 node
    120 ?        00:00:00 chrome <defunct>
    121 ?        00:00:00 chrome <defunct>
    155 ?        00:00:01 chrome <defunct>
    313 ?        00:00:00 chrome <defunct>
    314 ?        00:00:00 chrome <defunct>
    342 ?        00:00:02 chrome <defunct>
    520 ?        00:00:00 chrome <defunct>
    521 ?        00:00:00 chrome <defunct>
    554 ?        00:00:01 chrome <defunct>
    691 ?        00:00:00 chrome <defunct>
    692 ?        00:00:00 chrome <defunct>
    710 ?        00:00:01 chrome <defunct>
    901 ?        00:00:00 chrome <defunct>
    902 ?        00:00:00 chrome <defunct>
    936 ?        00:00:03 chrome <defunct>
   1103 ?        00:00:00 chrome <defunct>
   1104 ?        00:00:00 chrome <defunct>
   1138 ?        00:00:01 chrome <defunct>
   1283 ?        00:00:00 chrome <defunct>
   1284 ?        00:00:00 chrome <defunct>
   1304 ?        00:00:01 chrome <defunct>
   1484 ?        00:00:00 chrome <defunct>
   1485 ?        00:00:00 chrome <defunct>
   1519 ?        00:00:01 chrome <defunct>
   1696 ?        00:00:00 chrome <defunct>
   1697 ?        00:00:00 chrome <defunct>
   1731 ?        00:00:02 chrome <defunct>
   1912 ?        00:00:00 chrome <defunct>
   1913 ?        00:00:00 chrome <defunct>
   1941 ?        00:00:02 chrome <defunct>
   2090 ?        00:00:00 chrome <defunct>
   2091 ?        00:00:00 chrome <defunct>
   2124 ?        00:00:02 chrome <defunct>
   2268 ?        00:00:00 chrome <defunct>
   2269 ?        00:00:00 chrome <defunct>
   2303 ?        00:00:02 chrome <defunct>
   2432 ?        00:00:00 chrome <defunct>
   2433 ?        00:00:00 chrome <defunct>
   2454 ?        00:00:01 chrome <defunct>
   2609 ?        00:00:00 chrome <defunct>
   2610 ?        00:00:00 chrome <defunct>
   ..................
```
- I used `ps -e` command to show the all processes and found the tremendous defunct-ed chrome processes.

**Comment by prravda** [comment]
# Define the problem based on this metrics
## zombie processes are not managed by `init` process
```shell
root@hotdeal-scrapper:/app# ps -e
    PID TTY          TIME CMD
      1 ?        00:00:01 npm exec ts-nod
     16 ?        00:00:00 sh
     17 ?        00:47:26 node
    120 ?        00:00:00 chrome <defunct>
    121 ?        00:00:00 chrome <defunct>
    155 ?        00:00:01 chrome <defunct>
    313 ?        00:00:00 chrome <defunct>
```
- There are tremendous `defunct` processes (so called zombie processes) exist
- Headless browser process done it's job but it is not re-allocated into process pool so remained zombie process.
- For solving this problem, I used [`dumb-init`](https://github.com/Yelp/dumb-init) then it managed this situation properly  
## `BrowserContext` is cached continuously
- There is a object `BrowserContext` in this headless browser
- Basically this library is testing library, for fast test execution this object is cached
- But as time goes by, the data is accumulated into this object so it occupied more memory
- I explicitly close this browser context using `.close()` method

# Result


![Image](https://github.com/prravda/discord-hotdeal-scrapper-typescript/assets/64535895/9728c8a3-38ef-4ce8-a8ae-cdca7ce7852b)

- memory leak issue is resolved

**Comment by prravda** [comment]
# Additional update
- Changed DOM parsing library from `JSDOM` to `linkedom`
- It reduced 50% of memory usage for parsing html-string generated from buffer

**Comment by prravda** [comment]
# Result + @


![Image](https://github.com/prravda/discord-hotdeal-scrapper-typescript/assets/64535895/65f523f9-916d-4f1d-9a94-2e14853e5605)

Done.
